wit/tls.wit

```diff
@@ -0,0 +1,484 @@
+interface tls {
+    use wasi:io/streams@0.2.0.{input-stream, output-stream};
+    use wasi:io/poll@0.2.0.{pollable};
+
+    /// TLS protocol version.
+    ///
+    /// At the time of writing, these are the existing TLS versions:
+    /// - 0x0200: SSLv2 (Deprecated)
+    /// - 0x0300: SSLv3 (Deprecated)
+    /// - 0x0301: TLSv1.0 (Deprecated)
+    /// - 0x0302: TLSv1.1 (Deprecated)
+    /// - 0x0303: TLSv1.2
+    /// - 0x0304: TLSv1.3
+    ///
+    /// TODO: Want to use regular WIT `enum` type, but then adding a new protocol is backwards incompatible.
+    type protocol-version = u16;
+
+    /// TLS Cipher suite.
+    ///
+    /// These are maintained by IANA. Examples:
+    /// - 0x1301: TLS_AES_128_GCM_SHA256
+    /// - 0x1302: TLS_AES_256_GCM_SHA384
+    /// - 0xC030: TLS_ECDHE_RSA_WITH_AES_256_GCM_SHA384
+    /// - 0xC02B: TLS_ECDHE_ECDSA_WITH_AES_128_GCM_SHA256
+    /// - etc.
+    /// 
+    /// See: <https://www.iana.org/assignments/tls-parameters/tls-parameters.xhtml#tls-parameters-4>
+    type cipher-suite = u16;
+
+    /// Application-Layer Protocol Negotiation (ALPN) protocol ID.
+    ///
+    /// ALPN IDs are between 1 and 255 bytes long. Typically, they represent the
+    /// binary encoding of an ASCII string (e.g. `[0x68 0x32]` for `"h2"`),
+    /// though this is not required.
+    type alpn-id = list<u8>;
+
+    /// A X509 certificate chain; starting with the end-entity's certificate
+    /// followed by 0 or more intermediate certificates.
+    resource public-identity {
+        export-X509-chain: func() -> list<list<u8>>;
+    }
+
+    /// The combination of a private key with its public certificate(s).
+    /// The private key data can not be exported.
+    resource private-identity {
+        /// TODO: find a way to "preopen" these private-identity resources, so that the sensitive private key data never has to flow through the guest.
```

Comment:
In lieu of support for value imports, do we need something more complex than something like 
`import: static func(key-id: string) -> result<private-identity>;`
or potentially
`import: static func(key-id: string, x509-chain-id: string) -> result<private-identity>;`
?

Reply:
Something like that could probably work, yes.

My idea so far was to add an `identity-store` type, that can be used for both these "preopens" and also to access root certificates:
```wit
root-authorities: func() -> identity-store; // The TLS implementation's trust anchors.
user-identities: func() -> identity-store; // "preopens" per component instance

resource identity-store {
    all: func() -> list<identity>;
    
    // Optional utility methods:
    find-by-hostname: func(hostname: string) -> list<identity>;
    find-by-fingerprint: func(hash: string) -> option<identity>;
}
```